bst.py

```diff
@@ -131,17 +140,152 @@ def get_dot(self):
             )
         ))
 
+    def in_order(self):
+        return self._in_order(self.root)
+
+    def _in_order(self, leaf):
+        if leaf is None:
+            return
+        for value in self._in_order(leaf.left):
+            yield value
+        yield leaf.value
+        for value in self._in_order(leaf.right):
+            yield value
+
+    def pre_order(self):
+        return self._pre_order(self.root)
+
+    def _pre_order(self, leaf):
+        if leaf is None:
+            return
+        yield leaf.value
+        for value in self._pre_order(leaf.left):
+            yield value
+        for value in self._pre_order(leaf.right):
+            yield value
+
+    def post_order(self):
+        return self._post_order(self.root)
+
+    def _post_order(self, leaf):
+        if leaf is None:
+            return
+        for value in self._post_order(leaf.left):
+            yield value
+        for value in self._post_order(leaf.right):
+            yield value
+        yield leaf.value
+
+    def breadth_traversal(self):
+        x = deque()
+        x.append(self.root)
+        while x:
+            leaf = x.popleft()
+            yield leaf.value
+            if leaf.left:
+                x.append(leaf.left)
+            if leaf.right:
+                x.append(leaf.right)
+
+    def delete(self, val):
```

Comment:
This is a nice delete method, but it doesn't result in the tree re-balancing after a node is deleted.  That should happen.  So you need  a method similar to your `put` method below that does so.